tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc

```diff
@@ -270,9 +272,10 @@ StatusOr<tensorflow::AutotuneResult>
 GpuConvAlgorithmPicker::PickBestAlgorithmNoCacheCuda(
     const HloCustomCallInstruction& instr, se::DeviceMemoryAllocator* allocator,
     se::Stream* stream) {
-  XLA_SCOPED_LOGGING_TIMER(
-      absl::StrCat("GpuConvAlgorithmPicker::PickBestAlgorithmImpl for ",
-                   instr.ToString()));
+// Right now Redzone allocator is available in Cuda target only
+#if !TENSORFLOW_USE_ROCM
```

Comment:
Note: This is a hack. Redzone allocator won't compile in ROCm side. I'm putting it inside the scope of this function so as to not to put #if guard everywhere.

Reply:
Maybe implement Redzone allocator with `return Unimplemented()` or `CHECK()`? `#if` in XLA won't work nicely internally, so let's try our best avoiding them.

Reply:
I can do that for redzone allocator, but that is not the only place where macro `#if` is used. I'll comment out two other cases FYI. Do you mean that `#if` will not work at all in your master branch?

Reply:
> Do you mean that #if will not work at all in XLA in master branch?

It will work in XLA in master branch, but I can't speak for our internal version (which is synchronized with the opensource version). Summoning @Artem-B for more details.

Empirically, if you search `tensorflow/compiler/xla` for `GOOGLE_CUDA`, you will find only one file that uses it and that file may be a no-op, as many XLA/GPU applications don't define `GOOGLE_CUDA`.

Reply:
Had a brief discussion offline with @whchung. I am planning to do the following:

-  Create a common interface in `tensorflow/stream_executor/redzone_allocator`, and make both `ROCm` and `Cuda` inherit from this interface and have their own implementation. The difference between `Cuda` and `ROCm` is in constructor, that in Cuda side, `cuda::PtxCompilationOptions` is passed to constructor. `ROCm` does not have this field, because `PtxCompilation` is `Cuda` only.
- Needs to update signature of `se::cuda::PtxCompilationOptions PtxOptsFromConfig()` to be able to return a Status. So that ROCm implementation can return unimplemented
- The real problem is this [line](https://github.com/tensorflow/tensorflow/blob/ece727fd8c8d278a32357702fddb66caf40d4572/tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc#L294). `se::cuda::RedzoneAllocator` cannot be visible to `ROCm` anyways, it looks like I still need a macro here to remove the `PtxOptsFromConfig(hlo_module_config)` field.  
Since you have clarified that the problem is more in `GOOGLE_CUDA` undefined in XLA side, it looks like it doesn't prevent us from using `TENSORFLOW_USE_ROCM`, which is defined in `ROCm` compilation no matter what. Do you agree to use `TENSORFLOW_USE_ROCM` just for this place then?

Reply:
It looks like we want a portable type called something like `GpuAsmOpts`, and it wraps `cuda::PtxCompilationOptions`. We've done that to GpuVersion and I don't mind doing the same thing here.

With that, we may not need an interface type as `redzone_allocator.h` may be fully portable?

> Since you have clarified that the problem is more in GOOGLE_CUDA undefined in XLA side, it looks like it doesn't prevent us from using TENSORFLOW_USE_ROCM, which is defined in ROCm compilation no matter what.

It's not really "no matter what". If `--config=rocm` is not passed in, TENSORFLOW_USE_ROCM won't be defined. Internally XLA/GPU works without `--config=cuda`.

Reply:
Thanks for the clarification in `--config=rocm` macro.

Regarding to `redzone_allocator.h`, I agree that it can be shared across `ROCm` and `Cuda`. For `Cuda`side, I can create a typedef inside `ptxas_utils` to make `GpuAsmOpts` and `cuda::PtxCompilationOptions` equivalent. `ROCm` side can be all empty returns "unimplemented" for now.

Reply:
TL;DR; I agree with @timshen91 

I'm missing a lot of context here, so my usefulness is rather limited here. I can provide a bit of historic context and a dose of personal opinion. Feel free to skip the rant below.

`GOOGLE_CUDA` has outlived its usefulness long ago, IMO. Remaining uses are often of the `happens to work, until it does not` kind and are not a good example to follow. There's fair amount of internal history behind it, the underlying causes for its existences (first iteration of CUDA compiler which was not quite CUDA-compatible) are gone now, but the clean ups didn't quite catch up to it and new use cases seem to be cargo-culted faster than cleanups remove them. :-(

XLA does build w/o --config=cuda, because it does not have any CUDA code to compile. It does need CUDA runtime to execute JIT'ed code, but that's just plain old C API of CUDA runtime. One exception is NCCL which we do want to use and which is one of the few CUDA-related libraries XLA uses that does require CUDA compilation. I'm not sure what's the current status, but I believe we were leaning pre-compiling it (which would use --config=cuda) and using it from XLA as a regular library, so we remain relatively weakly tied to CUDA. 

In this sense HIP runtime should follow the same pattern, IMO. If you're not compiling any HIP code for XLA, we can (and should, I think) strive to avoid strongly tying XLA to HIP, too. 

Current implementation may need refactoring to accommodate both runtimes, but I think long term it will result in better code overall. Hopefully it will get us to find better solutions for the issues we run into, as opposed of working around them in the most expedient way. 

Long term I want to get rid of --config=cuda altogether and just build CUDA-enabled targets along with CPU-only ones, so a single build could produce both at the same time. HIP would hopefully fit within the same model. Then it would be up to the user what to include in the executable and up to the stream executor how to launch the kernel on particular accelerator. That, however, assumes that if the code is shared between different accelerators, the checks should be done at runtime, and not at compile time. It may be OK to use compile-time defines in ROCm-specific file and, e.g. make a choice between a stub vs full implementation, but in common code compile-time checks will eventually become a problem. We already had first glimpses of that dealing with CPU/GPU variants. With more than one GPU and, likely, other accelerator variants on the horizon thing will only get more interesting.

Reply:
Awesome thank you for the background and history @Artem-B @timshen91. Feel free to always share your vision up here, those kind of information will make sure we are aligned in the long run. I have shared your comments to the `TF-ROCm` team. Personally speaking, it makes me have a clear understanding of how much we want to avoid those macros :-)